packages/react/src/combobox/root/ComboboxRoot.tsx

```diff
@@ -0,0 +1,1135 @@
+  } = useOpenInteractionType(open);
+
+  const initialList = React.useMemo(() => {
+    if (virtualized) {
+      return Array.from({ length: flatFilteredItems.length }, () => null);
+    }
+    return [];
+  }, [virtualized, flatFilteredItems]);
+
+  const listRef = React.useRef<Array<HTMLElement | null>>(initialList);
+  const popupRef = React.useRef<HTMLDivElement | null>(null);
+  const valuesRef = React.useRef<Array<any>>([]);
+  const inputRef = React.useRef<HTMLInputElement | null>(null);
+  const keyboardActiveRef = React.useRef(true);
+  const allowActiveIndexSyncRef = React.useRef(true);
+  const hadInputClearRef = React.useRef(false);
+  const prevQueryRef = React.useRef(query);
+
+  const commitValidation = fieldControlValidation.commitValidation;
+
+  const updateValue = useEventCallback((nextValue: any) => {
+    clearErrors(name);
+    setDirty(nextValue !== validityData.initialValue);
+  });
+
+  const formValue = selectionMode === 'none' ? inputValue : selectedValue;
+
+  useField({
+    id,
+    commitValidation,
+    value: formValue,
+    controlRef: inputRef,
+    name,
+    getValue: () => formValue,
+  });
+
+  useIsoLayoutEffect(() => {
+    if (!virtualized) {
+      return;
+    }
+    // Drop stray nulls
+    listRef.current = initialList.slice();
+    valuesRef.current.length = initialList.length;
+  }, [initialList, virtualized]);
+
+  // Maintain a full value map for virtualized lists so offscreen selections and
+  // navigation use the correct indices even when DOM nodes aren't rendered.
+  useIsoLayoutEffect(() => {
+    if (!virtualized) {
+      return;
+    }
+    valuesRef.current = flatFilteredItems.slice();
+  }, [virtualized, flatFilteredItems]);
+
+  useIsoLayoutEffect(() => {
+    if (
+      prevQueryRef.current === query ||
+      !open ||
+      query === '' ||
+      query === String(defaultInputValue).toLocaleLowerCase()
+    ) {
+      return;
+    }
+
+    setQueryChangedAfterOpen(true);
+  }, [open, query, defaultInputValue]);
+
+  useIsoLayoutEffect(() => {
+    prevQueryRef.current = query;
+  }, [query]);
+
+  const prevValueRef = React.useRef(selectedValue);
+
+  useIsoLayoutEffect(() => {
+    if (prevValueRef.current === selectedValue) {
+      return;
+    }
+
+    clearErrors(name);
+    commitValidation?.(selectedValue, true);
+
+    if (validationMode === 'onChange') {
+      commitValidation?.(selectedValue);
+    }
+  }, [selectedValue, commitValidation, clearErrors, name, validationMode]);
+
+  useIsoLayoutEffect(() => {
+    const hasValue = multiple
+      ? Array.isArray(selectedValue) && selectedValue.length > 0
+      : selectedValue !== null && selectedValue !== undefined && selectedValue !== '';
+    setFilled(hasValue);
+    if (prevValueRef.current !== selectedValue) {
+      updateValue(selectedValue);
+    }
+  }, [setFilled, updateValue, selectedValue, multiple]);
+
+  useIsoLayoutEffect(() => {
+    prevValueRef.current = selectedValue;
+  }, [selectedValue]);
```

Comment:
For both `prevQueryRef` and `prevValueRef`, it would be preferable to encapsulate this behavior in a hook analogous to [useLatestRef](https://github.com/mui/base-ui/blob/master/packages/utils/src/useLatestRef.ts). It will be both more performant (no allocations), and have a better DX, its interface could look something like this:

```ts
const previousQuery = usePreviousValue(query)
```

Reply:
I struggled creating something that didn't mutate the ref in render. It seems like yours is doing that?

Reply:
Not sure what you mean by that. If you want to have the initial value be `undefined` rather than `value`, that's doable.